File: WindowsFormsApplicationLab2/WindowsFormsApplicationLaba2/ClassArray.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WindowsFormsApplicationLaba2
{
    class ClassArray<T> : IEnumerator<T>, IEnumerable<T>, IComparable<ClassArray<T>>
    {
        private Dictionary<int, T> places;
        private int maxCount;
        private T defaultValue;

        public ClassArray(int size, T defVal)
        {
            defaultValue = defVal;
            places = new Dictionary<int, T>();
            maxCount = size;
        }

        public static int operator +(ClassArray<T> p, T stone)
        {
            var isDiamond = stone is Diamond;
            if (p.places.Count == p.maxCount)
            {
                throw new ParkingOverflowException();
            }
            int index = p.places.Count;
            for (int i = 0; i < p.places.Count; i++)
            {
                if (p.CheckFreePlace(i))
                {
                    index = i;
                }
                if (stone.GetType() == p.places[i].GetType()) {
                    if (isDiamond)
                    {
                        if((stone as Diamond).Equals(p.places[i]))
                        {
                            throw new ParkingAlreadyHaveException();
                        }
                    }
                    else if((stone as Adamant).Equals(p.places[i]))
                    {
                        throw new ParkingAlreadyHaveException();
                    }
                }
            }
            if (index != p.places.Count)
            {
                p.places.Add(index,stone);
            }
            p.places.Add(p.places.Count, stone);
            return p.places.Count - 1;
        }

        public static T operator -(ClassArray<T> p, int index)
        {
            if (p.places.ContainsKey(index))
            {
                T stone = p.places[index];
                p.places.Remove(index);
                return stone;
            }
            throw new ParkingIndexOutOfRangeException();
        }

        private bool CheckFreePlace(int index)
        {
            return !places.ContainsKey(index);
        }

        public T this[int ind]
        {
            get
            {
                if (places.ContainsKey(ind))
                {
                    return places[ind];
                }
                return defaultValue;
            }
        }

        private int currentIndex;

        public T Current
        {
            get
            {
                return places[places.Keys.ToList()[currentIndex]];
            }
        }

        object IEnumerator.Current
        {
            get
            {
                return Current;
            }
        }

        public void Dispose() { }

        public bool MoveNext()
        {
            if (currentIndex + 1 >= places.Count)
            {
                Reset();
                return false;
            }
            currentIndex++;
            return true;
        }
        public void Reset()
        {
            currentIndex = -1;
        }

        public IEnumerator<T> GetEnumerator()
        {
            return this;
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        public int CompareTo(ClassArray<T> other)
        {
            if (this.Count() > other.Count())
            {
                return -1;
            }
            else if (this.Count() < other.Count())
            {
                return 1;
            }
            else
            {
                var thisKeys = places.Keys.ToList();
                var otherKeys = other.places.Keys.ToList();
                for (int i = 0; i < places.Count; ++i)
                {
                    if (places[thisKeys[i]] is Adamant && other.places[thisKeys[i]] is Diamond)
                    {
                        return 1;
                    }
                    if (places[thisKeys[i]] is Diamond && other.places[thisKeys[i]] is Adamant)
                    {
                        return -1;
                    }
                    if (places[thisKeys[i]] is Adamant && other.places[thisKeys[i]] is Adamant)
                    {
                        return (places[thisKeys[i]] is Adamant).CompareTo(other.places[thisKeys[i]] is Adamant);
                    }
                    if (places[thisKeys[i]] is Diamond && other.places[thisKeys[i]] is Diamond)
                    {
                        return (places[thisKeys[i]] is Diamond).CompareTo(other.places[thisKeys[i]] is Diamond);
                    }
                }
            }
            return 0;
        }
    }
}

File: TPlaba2_Java/src/ClassArray.java
import java.io.Serializable;
import java.util.Dictionary;
import java.util.Hashtable;
import java.util.Iterator;

public class ClassArray<T extends Stone> implements Serializable, Comparable<ClassArray<T>>, Iterable<T>, Iterator<T> {
	private T defaultValue;
	private int maxCount;
	private Dictionary<Integer, T> places;

	public ClassArray(int sizes, T defVal) {
		defaultValue = defVal;
		places = new Hashtable<Integer, T>();
		maxCount = sizes;

	}

	public static <T extends Stone> int plus(ClassArray<T> p, T stone)
			throws ParkingOverflowException, ParkingAlreadyHaveException {
		boolean isDiamond = stone instanceof Diamond;
		if (p.places.size() == p.maxCount) {
			throw new ParkingOverflowException();
		}
		int index = p.places.size();
		for (int i = 0; i < p.places.size(); i++) {
			if (p.CheckFreePlace(i))
				index = i;
			if (stone.getClass() == p.places.get(i).getClass()) {
				if (isDiamond) {
					if (((Diamond) stone).equals(p.places.get(i))) {
						throw new ParkingAlreadyHaveException();
					}
				} else if (((Adamant) stone).equals(p.places.get(i))) {
					throw new ParkingAlreadyHaveException();
				}
			}
		}
		if (index != p.places.size()) {
			p.places.put(index, stone);
			return index;
		}
		p.places.put(p.places.size(), stone);
		return p.places.size() - 1;
	}

	public static <T extends Stone> T minus(ClassArray<T> p, int index) throws ParkingIndexOutOfRangeException {
		if (((Hashtable<Integer, T>) p.places).containsKey(index)) {
			T stone = p.places.get(index);
			p.places.remove(index);
			return stone;
		}
		throw new ParkingIndexOutOfRangeException();
	}

	private boolean CheckFreePlace(int index) {
		return !((Hashtable<Integer, T>) places).containsKey(index);
	}

	public T getStone(int ind) {
		if (((Hashtable<Integer, T>) places).containsKey(ind)) {
			return places.get(ind);
		}
		return defaultValue;
	}

	@Override
	public Iterator<T> iterator() {
		return this;
	}

	private int currentIndex;

	@Override
	public boolean hasNext() {
		if (currentIndex + 1 >= places.size()) {
			currentIndex = -1;
			return false;
		}
		currentIndex++;
		return true;
	}

	@Override
	public T next() {
		return (T) places.get(currentIndex);
	}

	@Override
	public int compareTo(ClassArray<T> other) {
		if (places.size() > other.places.size())
			return -1;
		else if (places.size() < other.places.size())
			return 1;
		else {
			for (int i = 0; i < places.size(); i++) {
				if ((places.get(places.keys().nextElement()) instanceof Diamond)
						&& (other.places.get(other.places.keys().nextElement()) instanceof Diamond)) {
					return (((Adamant) places.get(places.keys().nextElement()))
							.compareTo(((Diamond) other.places.get(other.places.keys().nextElement()))));
				}
				if ((places.get(places.keys().nextElement()) instanceof Adamant)
						&& (other.places.get(other.places.keys().nextElement()) instanceof Diamond))
					return 1;
				if ((places.get(places.keys().nextElement()) instanceof Diamond)
						&& (other.places.get(other.places.keys().nextElement()) instanceof Adamant))
					return -1;
				if ((places.get(places.keys().nextElement()) instanceof Adamant)
						&& (other.places.get(other.places.keys().nextElement()) instanceof Adamant)) {
					return (((Adamant) places.get(places.keys().nextElement()))
							.compareTo(((Adamant) other.places.get(other.places.keys().nextElement()))));
				}

			}
		}
		return 0;
	}
}
